src/Agent.Plugins/PipelineArtifact/ArtifactProviderFactory.cs

```diff
@@ -0,0 +1,42 @@
+﻿using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.WebApi;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Agent.Sdk;
+using System.Threading;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    public class ArtifactProviderFactory
+    {
+        private readonly FileContainerProvider fileContainerProvider;
+        private readonly PipelineArtifactProvider pipelineArtifactProvider;
+
+        public ArtifactProviderFactory(AgentTaskPluginExecutionContext context, VssConnection connection, CallbackAppTraceSource tracer )
+        {
+            pipelineArtifactProvider = new PipelineArtifactProvider(context, connection, tracer);
+            fileContainerProvider = new FileContainerProvider(connection, tracer);
+        }
+
+        public IArtifactProvider GetProvider(BuildArtifact buildArtifact)
+        {
+            IArtifactProvider provider;
+            if (buildArtifact.Resource.Type == PipelineArtifactServer.PipelineArtifactTypeName)
```

Comment:
I think you can use a switch here so that
1) buildArtifact.Resource.Type will only be evaluated once
2) east to add another one